### PR TITLE
FSAI support with CUDA and HIP

### PR DESCRIPTION
**Pull request by victorapm**
Add CUDA and HIP support to FSAI, which can be also used as a preconditioner for BoomerAMG

PS: I left a few TODO items in `par_fsai_device.c` for better performance which I plan to implement on another PR.

### CODE REVIEW

**Review by rfalgout — APPROVED** [review]
Looks good to me.  Thanks!

**Comment by victorapm** [comment]
@ulrikeyang @osborn9 @liruipeng Can you take a look at this branch when you have time?

**Comment by osborn9** [comment]
Looks good! Nice work @victorapm! 

Minor issue - Seems like there are some issues after https://github.com/hypre-space/hypre/pull/739/commits/30efaf3da8c2a1d9cfb7ddbc627d532d7d7b7ed0 with some ILU functions.

**Comment by victorapm** [comment]
Thanks, Sarah! Maybe this is related to the new cuda-12 tests? Feel free to push any changes directly to this branch

File: src/parcsr_mv/par_csr_matop_device.c
Comment (ulrikeyang) [inline src/parcsr_mv/par_csr_matop_device.c:1]: Why is this now called twicw?

Comment (ulrikeyang) [reply src/parcsr_mv/par_csr_matop_device.c:1]: ignore this, this comment is in the wrong place.

Comment (ulrikeyang) [inline src/parcsr_mv/par_csr_matop_device.c:1]: Why is this here now called twice?

File: src/parcsr_mv/par_csr_matop_device.c [hunk 1]
```diff
@@ -1442,6 +1442,8 @@ hypre_ParCSRMatrixDropSmallEntriesDevice( hypre_ParCSRMatrix *A,
    hypre_TFree(tmp_j, HYPRE_MEMORY_DEVICE);
    hypre_GpuProfilingPopRange();
 
+   hypre_GpuProfilingPopRange();
```
Comment (victorapm) [inline src/parcsr_mv/par_csr_matop_device.c]: Hi @ulrikeyang, you mean this? Yes, second time is unnecessary, thanks for catching!

**Review by ulrikeyang — APPROVED** [review]
This looks fine, but I haven't really checked the GPU portions, since they are outside my area of expertise. Thanks

File: src/parcsr_mv/par_csr_matop_device.c (hunk 1 above)
Comment (victorapm) [reply src/parcsr_mv/par_csr_matop_device.c]: Fixed with d506b898c

**Review by osborn9 — APPROVED** [review]

File: src/parcsr_ls/par_relax.c [hunk 2]
```diff
@@ -762,6 +762,8 @@ hypre_BoomerAMGRelaxHybridGaussSeidel_core( hypre_ParCSRMatrix *A,
       v_buf_data = (HYPRE_Real *) hypre_ParCSRCommHandleSendDataBuffer(persistent_comm_handle);
       v_ext_data  = (HYPRE_Real *) hypre_ParCSRCommHandleRecvDataBuffer(persistent_comm_handle);
 #else
+      HYPRE_Int num_cols_offd = hypre_CSRMatrixNumCols(A_offd);
```
Comment (liruipeng) [inline src/parcsr_ls/par_relax.c]: Why defining again? Is it different from the one at line 714?

Comment (victorapm) [reply src/parcsr_ls/par_relax.c]: Agreed! Let me remove this redundant line

Comment (victorapm) [reply src/parcsr_ls/par_relax.c]: Fixed with 6df6f9b

File: src/parcsr_ls/par_relax.c [hunk 3]
```diff
@@ -785,6 +786,7 @@ hypre_BoomerAMGRelaxHybridGaussSeidel_core( hypre_ParCSRMatrix *A,
 #if defined(HYPRE_USING_PERSISTENT_COMM)
       hypre_ParCSRPersistentCommHandleStart(persistent_comm_handle, HYPRE_MEMORY_HOST, v_buf_data);
 #else
+      hypre_ParCSRCommHandle *comm_handle;
```
Comment (liruipeng) [inline src/parcsr_ls/par_relax.c]: Why do we need to define another `comm_handle` here?

Comment (victorapm) [reply src/parcsr_ls/par_relax.c]: This is not needed, thanks for catching!

Comment (victorapm) [reply src/parcsr_ls/par_relax.c]: Fixed with [e082daf](https://github.com/hypre-space/hypre/pull/739/commits/e082dafda6a2ef544eda2fca27d62141d433872b)

File: src/utilities/device_utils.c [hunk 4]
```diff
@@ -526,6 +526,61 @@ hypre_dim3(HYPRE_Int x, HYPRE_Int y, HYPRE_Int z)
...
+
+   return hypre_error_flag;
+}
+
+/*--------------------------------------------------------------------
+ * hypreDevice_ComplexArrayToArrayOfPtrs
+ *--------------------------------------------------------------------*/
+
+HYPRE_Int
+hypreDevice_ComplexArrayToArrayOfPtrs(HYPRE_Int       n,
```
Comment (liruipeng) [inline src/utilities/device_utils.c:576]: The name of this function is a bit confusing. I understood what it is after reading the code. So, it basically just gets the addresses of the elements of `data` and save them in another array.

File: src/utilities/device_utils.h [hunk 5]
```diff
@@ -13,9 +13,11 @@
 /* Data types depending on GPU architecture */
 #if defined(HYPRE_USING_CUDA) || defined(HYPRE_USING_SYCL)
 typedef hypre_uint          hypre_mask;
+#define hypre_mask_one      1U
 
 #elif defined(HYPRE_USING_HIP)
 typedef hypre_ulonglongint  hypre_mask;
+#define hypre_mask_one      1ULL
```
Comment (liruipeng) [inline src/utilities/device_utils.h:20]: better to move this together with `HYPRE_WARP_FULL_MASK` and also change to all capitalized letters?

**Review by liruipeng — APPROVED** [review]
Looks good to me. A lot of nice work. Thanks @victorapm

File: src/utilities/device_utils.c (hunk 4 above)
Comment (victorapm) [reply src/utilities/device_utils.c:576]: Exactly. It forms a double pointer. This is expected by some vendor math libraries... Do you have any suggestions for a different name?

File: src/utilities/device_utils.h (hunk 5 above)
Comment (victorapm) [reply src/utilities/device_utils.h:20]: Yes, agreed

**Comment by victorapm** [comment]
Thanks for reviewing! I'll run all the regression tests again before merging this into master

File: src/utilities/device_utils.c (hunk 4 above)
Comment (liruipeng) [reply src/utilities/device_utils.c:576]: Something like "GetArrayElemPtrs"? The current one reads like we are converting `Complex` to `Complex *`. Thanks!

Comment (victorapm) [reply src/utilities/device_utils.c:576]: Good suggestion! What about `ComplexArrayGetArrayOfPtrs`? I would like to have `ComplexArray` at the beginning because we might want to do the same op for integers as well, so we would be able to distinguish with something like `IntArrayGetArrayOfPtrs`? Note the output is an array of pointers, hence the name `ArrayOfPtrs`

Comment (liruipeng) [reply src/utilities/device_utils.c:576]: Sounds good to me. Thank you!
